packages/sanity/src/core/store/_legacy/document/utils/eventChainUtils.ts

```diff
@@ -0,0 +1,49 @@
+import {partition} from 'lodash'
+
+import {type MutationEvent} from '../types'
+
+export function discardChainTo(events: MutationEvent[], revision: string) {
+  const sortedChain = linkedSort(events)
+  const revisionIndex = sortedChain.findIndex((event) => event.resultRev === revision)
+
+  return split(events, revisionIndex)
+}
+
+function split<T>(array: T[], index: number) {
+  if (index < 0) {
+    return [[], array]
+  }
+  return [array.slice(0, index), array.slice(index + 1)]
+}
+
+export function linkedSort<T extends {previousRev: string; resultRev: string}>(events: T[]) {
+  const parents: Record<string, T | undefined> = {}
+
+  events.forEach((event) => {
+    parents[event.resultRev] = events.find((other) => other.resultRev === event.previousRev)
```

Comment:
`events.forEach` with  a nested `events.find` can be slow for large amounts of events. i think we only have at most 10 events so this may not need to be refactored though.